tools/shared.py

```diff
@@ -1824,6 +1824,9 @@ def llvm_backend_args():
     # and https://bugs.llvm.org/show_bug.cgi?id=39488
     args += ['-disable-lsr']
 
+    if Settings.REFERENCE_TYPES:
+      args += ['-mattr=reference-types']
```

Comment:
Can we just avoid this new setting and have the user pass `-mattr=reference-types` on the command line directly?

Reply:
I suppose we may also need a corresponding flag to binaryen, which would require having a global flag or are you proposing we could parse out the mattr from the `emcc` line and figure out what to tell binaryen?

Reply:
Not really. I have not found a way to percolate this information down to `wasm-ld` when invoked through `emcc` and everything else seems to also make use of a new setting.

Reply:
But this is a compiler flag not a linker flag right?   Are you thinking of the LTO case? 

I suppose for the LTO case it might make sense, but the way to do this with the normal clang toolchain would be using `-Wl` to pass flags the linker.   In this case i guess it would be `-Wl,-mllvm,-mattr=reference-types`, which I agree is kind of a mouthful.   I wonder how the clang driver handles this, does it pass codegen flags to the the linker when doing LTO?

Reply:
Just passing `-mreference-types` will automatically propagate the feature information to the linker and to Binaryen using [the target features section](https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md#target-features-section), so there should not be any need for special modes or flags.

Reply:
> But this is a compiler flag not a linker flag right? Are you thinking of the LTO case?

The LLVM backend in `wasm-ld` needs to have set that attribute. 

> Just passing `-mreference-types` will automatically propagate the feature information to the linker and to Binaryen using [the target features section](https://github.com/WebAssembly/tool-conventions/blob/master/Linking.md#target-features-section), so there should not be any need for special modes or flags.

I don't see `-matomics` or `-mreference-types` propagate do the invocation of `wasm-ld`:

```
emcc -g2 -v -s SIDE_MODULE=1 -matomics -mreference-types -s WASM=1 /data/vchuravy/llvm-project/llvm/test/CodeGen/WebAssembly/anyref.ll -o anyref.wasm
shared:INFO: (Emscripten: Running sanity checks)
 "/data/vchuravy/llvm-build/bin/wasm-ld" -o /tmp/emscripten_temp_azWLWt/anyref.wasm --allow-undefined --lto-O0 --whole-archive /tmp/emscripten_temp_azWLWt/anyref_0.o -L/data/vchuravy/emsdk/emscripten/incoming/system/local/lib -L/data/vchuravy/emsdk/emscripten/incoming/system/lib -L/home/vchuravy/.emscripten_cache/wasm-obj-pic --no-whole-archive --import-memory --import-table -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --no-gc-sections --export-dynamic --export-all --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location --export fflush -shared
```

Reply:
Actually the target features section is correctly populated and parsed. Since the below works

```
llc -mattr=reference-types --filetype=obj anyref.ll -o anyref.o
wasm-ld --allow-undefined --no-entry --export-dynamic -o anyref.wasm
wasm-dis anyref.wasm
```

The issue is more that invoking `emcc` on the .ll file directly leads to a crash, since `emcc` does not invoke `llc`, but uses `llvm-as`,

```
emcc:DEBUG: assembling assembly file: /data/vchuravy/llvm-project/llvm/test/CodeGen/WebAssembly/anyref.ll
shared:DEBUG: successfully executed /data/vchuravy/llvm-build/bin/llvm-as /data/vchuravy/llvm-project/llvm/test/CodeGen/WebAssembly/anyref.ll -o /tmp/emscripten_temp/anyref_0.o
emcc:DEBUG: linking: ['/tmp/emscripten_temp/anyref_0.o', '-L/data/vchuravy/emsdk/emscripten/incoming/system/local/lib', '-L/data/vchuravy/emsdk/emscripten/incoming/system/lib', '-L/home/vchuravy/.emscripten_cache/wasm-obj-pic']
LLVM ERROR: Cannot select: 0x5606c189b6e8: anyref,ch = WebAssemblyISD::CALL1 0x5606c182a4e8, 0x5606c189b680, 0x5606c189b548
  0x5606c189b680: i32 = WebAssemblyISD::Wrapper TargetGlobalAddress:i32<i8 addrspace(256)* (i8 addrspace(256)*)* @test> 0
    0x5606c189b618: i32 = TargetGlobalAddress<i8 addrspace(256)* (i8 addrspace(256)*)* @test> 0
  0x5606c189b548: anyref = WebAssemblyISD::ARGUMENT TargetConstant:i32<0>
    0x5606c189b4e0: i32 = TargetConstant<0>
In function: call_test
shared:ERROR: '/data/vchuravy/llvm-build/bin/wasm-ld -o /tmp/emscripten_temp/anyref.wasm --allow-undefined --lto-O0 --whole-archive /tmp/emscripten_temp/anyref_0.o -L/data/vchuravy/emsdk/emscripten/incoming/system/local/lib -L/data/vchuravy/emsdk/emscripten/incoming/system/lib -L/home/vchuravy/.emscripten_cache/wasm-obj-pic --no-whole-archive --import-memory --import-table -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --no-gc-sections --export-dynamic --export-all --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location --export fflush -shared' failed (1)
cache:DEBUG: PID 7437 released multiprocess file lock to Emscripten cache at /home/vchuravy/.emscripten_cache/wasm-obj-pic
```

Reply:
Hmm, it seems to me that we should be using `clang` as the driver for compiling .ll files rather than `llvm-as` so the normal clang target feature flags can be used. @kripken, @sbc100 what do you think of that? Running emcc on .ll files is uncommon enough in normal usage that I definitely think it's not worth adding an Emscripten setting for.

Reply:
Okay if I add a `target-features` function attribute then the `emcc` pipeline can also consume that file successfully.

Reply:
@sbc100 the problem was that hand-written .ll files didn’t not necessarily have that target feature information. This PR gave a way to provide the target feature info as a flag in that case, but the better fix is to just add the info to the .ll file. So this PR is not necessary.